### PR TITLE
Created backend server foundation

### PR DESCRIPTION
**Pull request by apfohl**
- Express
- InversifyJS
- WebPack building
- Services: ApplicationService, SessionService
- Controllers: ExpensesController, SessionController
- Middlewares: LoggerMiddleware
- Models: Expense, Session, User
- Helpers: TokenGenerator

### CODE REVIEW

**Comment by apfohl** [comment]
`webpack.config.js` is missing. I'll add it.

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,7 @@
+node_modules
+dist
+*.log
+*.js
```
Comment (Kharonus) [inline .gitignore]: seems odd to exclude all js files. Are we sure we need none? also not for configuring any dependencies?

File: src/application.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+// Controllers
+import './controllers/expenses.controller';
+import './controllers/session.controller';
+
+// InversifyJS Container
+const container = new Container();
+
+// Bind application service
+const applicationService: ApplicationService = new ApplicationService();
+container.bind<ApplicationService>(ApplicationService.name).toConstantValue(applicationService);
```
Comment (Kharonus) [inline src/application.ts]: Can you check if inversify gives you the possibilty to create singletons by itself.

See https://github.com/inversify/InversifyJS/issues/560

File: src/helpers/token-generator.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+            return token;
+        }
+    }
+
+    private static generateToken(length: number = 6): string {
+        let token = '';
+
+        for (let i = 0; i < length - 1; i++) {
+            token += TokenGenerator.alphabet.charAt(Math.floor(Math.random() * TokenGenerator.alphabet.length));
```
Comment (Kharonus) [inline src/helpers/token-generator.ts]: instead of charAt() you can simply use []

File: src/helpers/token-generator.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
+export class TokenGenerator {
+
+    private static alphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
+    private static usedTokens: Array<string> = [];
+
+    public static newToken(): string {
+        while (true) {
+            const token: string = this.generateToken();
+
+            if (this.usedTokens.indexOf(token) > -1) {
```
Comment (Kharonus) [inline src/helpers/token-generator.ts]: using typescript 3 we can use es6 function include -> if (this.usedTokens.include(token)) {}

File: src/controllers/session.controller.ts [hunk 5]
```diff
@@ -0,0 +1,25 @@
+import { Response } from 'express';
+import { inject } from 'inversify';
+import { controller, httpPost, requestBody, response } from 'inversify-express-utils';
+
+import { LoggerMiddleware } from '../middleware/logger.middleware';
+import { User } from '../models/user.model';
+import { SessionService } from '../services/session.service';
+
+@controller('/authentication', LoggerMiddleware)
+export class AuthenticationController {
```
Comment (Kharonus) [inline src/controllers/session.controller.ts]: This should be a SessionController

File: src/services/application.service.ts [hunk 6]
```diff
@@ -0,0 +1,19 @@
+import { injectable } from 'inversify';
+import { Server } from 'http';
+
+@injectable()
```
Comment (Kharonus) [inline src/services/application.service.ts]: Is it possible to use here the same technique as in the SessionService

File: .gitignore (hunk 1 above)
Comment (apfohl) [reply .gitignore]: I `git add -f` them right now. But we can write exceptions into `.gitignore`-

File: src/services/session.service.ts [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+@fluentProvide(SessionService).inSingletonScope().done()
+export class SessionService {
+
+    private sessions: Array<Session> = [];
+
+    public createSession(user: User): void {
+        const token = TokenGenerator.newToken();
+        const session = new Session(user, token);
+
+        this.sessions.push(session);
```
[the remark this answers is not in the document]
Comment (apfohl) [reply src/services/session.service.ts]: This is just a placeholder for the real session implementation.

File: src/application.ts (hunk 2 above)
Comment (apfohl) [reply src/application.ts]: It does indeed.

**Pull request merged**